boot/bootutil/src/loader.c

```diff
@@ -793,58 +855,122 @@ boot_copy_sector(int flash_area_id_src, int flash_area_id_dst,
 static void
 boot_swap_sectors(int idx, uint32_t sz, struct boot_status *bs)
 {
+    const struct flash_area *fap;
     uint32_t copy_sz;
+    uint32_t trailer_sz;
     uint32_t img_off;
+    uint32_t scratch_trailer_off;
+    struct boot_swap_state swap_state;
     int rc;
 
     /* Calculate offset from start of image area. */
     img_off = boot_data.imgs[0].sectors[idx].fa_off -
               boot_data.imgs[0].sectors[0].fa_off;
 
+    copy_sz = sz;
+
+    trailer_sz = boot_slots_trailer_sz(boot_data.write_sz);
+    if (boot_data.imgs[0].sectors[idx].fa_off + sz >
+        boot_data.imgs[0].sectors[boot_data.imgs[0].num_sectors - 1].fa_off) {
+        copy_sz -= trailer_sz;
+    }
+
+    bs->use_scratch = (bs->idx == 0 && copy_sz != sz);
+
     if (bs->state == 0) {
         rc = boot_erase_sector(FLASH_AREA_IMAGE_SCRATCH, 0, sz);
         assert(rc == 0);
 
         rc = boot_copy_sector(FLASH_AREA_IMAGE_1, FLASH_AREA_IMAGE_SCRATCH,
-                              img_off, 0, sz);
+                              img_off, 0, copy_sz);
         assert(rc == 0);
 
+        if (bs->idx == 0) {
```

Comment:
I believe this has problems, but perhaps I'm missing something.

(By the way, the commit logs in this series are a little terse -- do you have any pointers to more detailed discussion explaining these changes? It would help me understand what's going on a bit better.)

Anyway, this if statement is going to at least break some targets, e.g. the STM32F401RE-based 96Boards Carbon, which is a Zephyr target. I'm not sure how it's meant to work in general, but it looks to me that there are other issues too.

On STM32F401RE, the sectors have variable sizes. Here is the layout:

| Sector | Start address | End address | Size (KiB) | Size (B) | User           |
|--------|---------------|-------------|------------|----------|----------------|
| 0      | 0x08000000    | 0x08003FFF  | 16         | 0x4000   | mcuboot        |
| 1      | 0x08004000    | 0x08007FFF  | 16         | 0x4000   | mcuboot        |
| 2      | 0x08008000    | 0x0800BFFF  | 16         | 0x4000   | None           |
| 3      | 0x0800C000    | 0x0800FFFF  | 16         | 0x4000   | FOTA app state |
| 4      | 0x08010000    | 0x0801FFFF  | 64         | 0x10000  | None           |
| 5      | 0x08020000    | 0x0803FFFF  | 128        | 0x20000  | Image 1        |
| 6      | 0x08040000    | 0x0805FFFF  | 128        | 0x20000  | Image 2        |
| 7      | 0x08060000    | 0x0807FFFF  | 128        | 0x20000  | Scratch        |

As you can see, the images and scratch each take up exactly one of the three 128KB size sectors at the end of the device flash.

Here's what will happen during a copy:

1. line 881, boot_erase_sector() erases scratch (nothing is lost)
2. line 884, boot_copy_sector() copies image 1 into scratch (nothing is lost)
3. line 896, boot_erase_last_sector_by_id() erases the entire image 0 sector (**data loss; mcuboot cannot revert if image 1 is bad**)

I'm not sure what is supposed to be happening in the call to boot_status_init_by_id() on line 899. It looks like what will happen is:

- Read image 1 trailer's swap state (magic, copy_done, image_ok)
- If image 1's image_ok is set, write image_ok to image 0 trailer
- Write image 0 trailer magic

It seems like the idea is to set up the image 0 trailer to be at least a test swap from image 1, and maybe a permanent swap if image 1 has the image_ok flag set.

But that doesn't make sense to me.

- If the trailer is smaller than the last sector in image 0, **some image 0 data can get permanently erased on line 896**. This is true even on targets where the sector size is smaller than the image area, **not just on 96Boards Carbon**.
- If the trailer is larger than the last sector in image 0, then not enough of flash has been erased in line 896, and **mcuboot might later try to write data in image 0 it has not erased yet**, e.g. writing the magic bytes during the call to boot_status_init_by_id(), depending on the sector size.
- If the trailer is exactly the same size as the last sector of image 0, then things seem like they may be OK, but that's **not a safe assumption** (e.g. it's not the case on any Zephyr target).

What's going on here? Am I missing something?